### ISSUE TITLE
CME and visual problem when switching back to Home tab

### ISSUE THREAD

**Issue by NPException** [oakes/Lightmod#3]
TL;DR:
1. Java 9 complains about illegal reflective access
2. ConcurrentModificationException when switching back to Home tab after the chat project started completely
3. Background in Home area forgot to rerender

_I'm using the .jar Version of LightMod 1.0.2, with Java 9.0.1 on Windows. (But the issue also happend with LightMod 1.0.1 and Java 1.8)_

I can reproduce the issue every time with the following steps:

Step 1) Start LightMod. (in my case from the command line to see stdout)

As a side note, Java 9 complains about illegal reflective access at this point. This is my entire stdout at the point when the LightMod GUI becomes visible:
```
E:\downloads\lightmod>java -Xms2g -Xmx4g -jar .\Lightmod-1.0.2.jar
2017-11-15 16:58:28.098:INFO::JavaFX Application Thread: Logging initialized @1268ms
2017-11-15 16:58:29.297:INFO:oejs.Server:main: jetty-9.2.z-SNAPSHOT
2017-11-15 16:58:29.314:INFO:oejs.ServerConnector:main: Started ServerConnector@51e2adc7{HTTP/1.1}{0.0.0.0:56001}
2017-11-15 16:58:29.316:INFO:oejs.Server:main: Started @2485ms
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by clojure.lang.Reflector (file:/E:/downloads/lightmod/Lightmod-1.0.2.jar) to method com.sun.javafx.collections.VetoableListDecorator.add(java.lang.Object)
WARNING: Please consider reporting this to the maintainers of clojure.lang.Reflector
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
```

Step 2) Click on the "chat" project tab and wait until the server is started, and the chat is useable

Step 3) Click on the "Home" tab.

LightMod does not seem to react to that click, but the following Exception is printed to the console:

```
Exception in thread "JavaFX Application Thread" java.util.ConcurrentModificationException
        at java.base/java.util.HashMap$HashIterator.nextNode(Unknown Source)
        at java.base/java.util.HashMap$KeyIterator.next(Unknown Source)
        at java.base/java.util.Collections$UnmodifiableCollection$1.next(Unknown Source)
        at org.httpkit.server.HttpServer.stop(HttpServer.java:334)
        at org.httpkit.server$run_server$stop_server__7318.doInvoke(server.clj:56)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:29)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at lightmod.app$stop_server_BANG_.invokeStatic(app.clj:200)
        at lightmod.app$stop_server_BANG_.invoke(app.clj:196)
        at lightmod.app$stop_app_BANG_.invokeStatic(app.clj:320)
        at lightmod.app$stop_app_BANG_.invoke(app.clj:319)
        at lightmod.ui$create_tab$reify__7954.handle(ui.clj:69)
        at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(Unknown Source)
        at javafx.base/javafx.event.Event.fireEvent(Unknown Source)
        at javafx.controls/javafx.scene.control.Tab$1.invalidated(Unknown Source)
        at javafx.base/javafx.beans.property.BooleanPropertyBase.markInvalid(Unknown Source)
        at javafx.base/javafx.beans.property.BooleanPropertyBase.set(Unknown Source)
        at javafx.controls/javafx.scene.control.Tab.setSelected(Unknown Source)
        at javafx.controls/javafx.scene.control.TabPane$TabPaneSelectionModel.select(Unknown Source)
        at javafx.controls/javafx.scene.control.TabPane$TabPaneSelectionModel.select(Unknown Source)
        at javafx.controls/javafx.scene.control.TabPane$TabPaneSelectionModel.select(Unknown Source)
        at javafx.controls/com.sun.javafx.scene.control.behavior.TabPaneBehavior.selectTab(Unknown Source)
        at javafx.controls/javafx.scene.control.skin.TabPaneSkin$TabHeaderSkin$5.handle(Unknown Source)
        at javafx.controls/javafx.scene.control.skin.TabPaneSkin$TabHeaderSkin$5.handle(Unknown Source)
        at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(Unknown Source)
        at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(Unknown Source)
        at javafx.base/javafx.event.Event.fireEvent(Unknown Source)
        at javafx.graphics/javafx.scene.Scene$MouseHandler.process(Unknown Source)
        at javafx.graphics/javafx.scene.Scene$MouseHandler.access$1300(Unknown Source)
        at javafx.graphics/javafx.scene.Scene.processMouseEvent(Unknown Source)
        at javafx.graphics/javafx.scene.Scene$ScenePeerListener.mouseEvent(Unknown Source)
        at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(Unknown Source)
        at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(Unknown Source)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$2(Unknown Source)
        at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(Unknown Source)
        at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(Unknown Source)
        at javafx.graphics/com.sun.glass.ui.View.handleMouseEvent(Unknown Source)
        at javafx.graphics/com.sun.glass.ui.
```

Step 4) Click on the "Home" tab a second time:
![image](https://user-images.githubusercontent.com/6052580/32846743-5982ab56-ca28-11e7-9f3a-ab43a81457be.png)

The visual of the "chat" project is the background of the "Home" area now. Only the regular "Home" area buttons are useable, so I suspect it just missed to render the gray background color again?

If I can provide you with any more information, let me know. :)

**Comment by oakes** [comment]
Wow interesting! I definitely haven't seen this before, but the stack trace will definitely help. Does this happen with the installer version, or just the jar?

**Comment by NPException** [comment]
I can reliably reproduce it with the installer version as well. Both, 1.0.1 and 1.0.2.

Side note: I just noticed that on Step 3 in my description, the stacktrace also appears in the "Server logs" section.

**Comment by oakes** [comment]
OK I'll be releasing a new version soon that will hopefully address this.

**Comment by oakes** [comment]
I released 1.0.3 that should catch that exception. Please let me know if you still notice it. Thanks again! This has been extremely helpful for me.

**Comment by NPException** [comment]
@oakes , I'm afraid you need to reopen the issue.

I just tested with 1.0.4, and can reproduce the exact same behavior. Only some line numbers in the Stacktrace changed:

```
Exception in thread "JavaFX Application Thread" java.util.ConcurrentModificationException
        at java.base/java.util.HashMap$HashIterator.nextNode(Unknown Source)
        at java.base/java.util.HashMap$KeyIterator.next(Unknown Source)
        at java.base/java.util.Collections$UnmodifiableCollection$1.next(Unknown Source)
        at org.httpkit.server.HttpServer.stop(HttpServer.java:334)
        at org.httpkit.server$run_server$stop_server__7318.doInvoke(server.clj:56)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.lang.AFunction$1.doInvoke(AFunction.java:29)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at lightmod.app$stop_server_BANG_.invokeStatic(app.clj:200)
        at lightmod.app$stop_server_BANG_.invoke(app.clj:196)
        at lightmod.app$stop_app_BANG_.invokeStatic(app.clj:320)
        at lightmod.app$stop_app_BANG_.invoke(app.clj:319)
        at lightmod.ui$create_tab$reify__7954.handle(ui.clj:69)
        at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(Unknown Source)
```

**Comment by oakes** [comment]
Gotcha, thank you for letting me know; I'll look into this later today.

**Comment by NPException** [comment]
This may be a bug in http-kit: https://github.com/http-kit/http-kit/issues/355

**Comment by oakes** [comment]
The weird thing is that I [added a try/catch](https://github.com/oakes/Lightmod/blob/master/src/clj/lightmod/app.clj#L201), but your stack trace says the exception is happening on line 200 which is where it says `(when server`. I wonder if moving the `(when server` inside of the try/catch would allow it to catch the exception. If you happen to have the repo cloned, you can try that out, but either way I'll try to investigate more later.

**Comment by NPException** [comment]
When I start it from the source using boot, the stacktrace points to a different line number: 203
When trying around a bit more, it looks like the Exception could occur on the `(server)` and `(reload-stop-fn)` calls. (I may be wrong though)

After I wrapped the latter one in a try-catch as well, the exception and the graphical glitch are gone.

**Comment by oakes** [comment]
Aha! The reload server also uses http-kit, I completely forgot about that! I'll cut a new release shortly. Once you get a chance to try it and confirm it fixes it, we can close this. I really appreciate the help!

**Comment by NPException** [comment]
Glad I can help! :)

**Comment by oakes** [comment]
I'm still building the windows/linux installers but you can download the jar file now if you'd like:

https://github.com/oakes/Lightmod/releases/tag/1.0.5

**Comment by NPException** [comment]
Everything's fine with 1.0.5 👍

**Comment by oakes** [comment]
Sweet! Thanks again, and of course I'm all ears if you run into other stuff.
